libbeat/registry/registry.go

```diff
@@ -0,0 +1,92 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package registry
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/registry/backend"
+)
+
+// Registry manages multiple key-value stores.
+// When working with a registry, one must access a store. Depending on backend
+// a store can be an index, a table, or a directory. All access to a store is
+// handled by transaction.
+type Registry struct {
+	backend backend.Registry
+
+	mu     sync.Mutex
+	active map[string]*sharedStore // active/open stores
+	wg     sync.WaitGroup
+}
+
+// Key is the type used for all keys in a store.
+type Key []byte
+
+// ValueDecoder is used to decode retrieved from an actual store.  A
+// ValueDecoder instance is valid for the lifetime of the transaction only.
+type ValueDecoder interface {
+	Decode(to interface{}) error
+}
+
+// NewRegistry creates a new Registry with a configured backend.
+func NewRegistry(backend backend.Registry) *Registry {
+	return &Registry{
+		backend: backend,
+		active:  map[string]*sharedStore{},
+	}
+}
+
+// Close closes the backend storage. Close blocks until all stores in use are closed.
+func (r *Registry) Close() error {
+	r.wg.Wait() // wait for all stores being closed
+	return r.backend.Close()
+}
+
+// Get opens a shared store. A store is closed and released only after all it's
+// users have closed the store.
+func (r *Registry) Get(name string) (*Store, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	shared := r.active[name]
+	if shared == nil {
+		backend, err := r.backend.Access(name)
+		if err != nil {
+			return nil, err
+		}
+
+		shared = newSharedStore(r, name, backend)
+		r.active[name] = shared
+		r.wg.Add(1)
+	} else {
+		shared.refCount.Inc()
```

Comment:
`shared.refCount.Inc()` looks like implementations details, suggestion is to hide the change behind a method.

My really old objective background make me things of: 

```
shared.Retain() // shared.RefInc()
```

Reply:
haha... yeah, I can name them Retain and Release if it makes you happy :)